blockchain/merkle.go

```diff
@@ -141,3 +140,91 @@ func BuildMerkleTreeStore(transactions []*btcutil.Tx, witness bool) []*wire.ShaH
 
 	return merkles
 }
+
+// ExtractWitnessCommitment attempts to locate, and return the witness
+// commitment for a block. The witness commitment is of the form:
+// SHA256(witness root || witness nonce).The function additionally returns a
+// boolean indicating if the witness root was located within any of the txOut's
+// in the passed transaction. The witness commitment is stored as the data push
+// for an OP_RETURN with special magic bytes to aide in location.
+func ExtractWitnessCommitment(tx *btcutil.Tx) ([]byte, bool) {
+	var witnessCommitment []byte
+
+	// The witness commitment *must* be located within one of the coinbase
+	// transaction's outputs.
+	if !IsCoinBase(tx) {
+		return witnessCommitment, false
+	}
+
+	msgTx := tx.MsgTx()
+	witFound := false
+	for i := len(msgTx.TxOut) - 1; i >= 0; i-- {
+		// The public key script that contains the witness commitment
+		// must shared a prefix with the WitnessMagicBytes, and be at
+		// least 38 bytes.
+		if bytes.HasPrefix(msgTx.TxOut[i].PkScript, WitnessMagicBytes) &&
```

Comment:
I've modified the code with your suggestions, including adding some new constants for the magic offsets, and lengths. 

As you've probably seen within `ValidateWitnessCommitment`, if the coinbase doesn't have something which looks like a witness commitment within any of its outputs, then if the transactions within the block contain _any_ witness data, then the block itself is invalid. Therefore it's necessary to also return if the commitment was found or not. Hmm, but I guess the function could simply return `nil` if the commitment was not found.